File: frontend/src/services/translationService.ts
import { API_BASE } from '../utils/constants';

export async function translateText(text: string, source: string, target: string): Promise<string> {
  try {
    const res = await fetch(`${API_BASE}/translate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text, source, target }),
    });
    if (!res.ok) throw new Error('Translation failed');
    const data = await res.json();
    return data.translatedText;
  } catch {
    // Fallback to MyMemory API
    return translateWithMyMemory(text, source, target);
  }
}

async function translateWithMyMemory(text: string, source: string, target: string): Promise<string> {
  const url = `https://api.mymemory.translated.net/get?q=${encodeURIComponent(text)}&langpair=${source}|${target}`;
  const res = await fetch(url);
  if (!res.ok) return text;
  const data = await res.json();
  return data.responseData?.translatedText ?? text;
}

export async function translateBatch(texts: string[], source: string, target: string): Promise<string[]> {
  try {
    const res = await fetch(`${API_BASE}/translate-batch`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ texts, source, target }),
    });
    if (!res.ok) throw new Error('Batch translation failed');
    const data = await res.json();
    return data.translations;
  } catch {
    // Fallback: translate one-by-one
    return Promise.all(texts.map(t => translateText(t, source, target)));
  }
}

File: backend/services/translation_service.py
import httpx
from config import settings


async def translate(text: str, source: str, target: str) -> str:
    async with httpx.AsyncClient(timeout=10.0) as client:
        try:
            res = await client.post(
                f"{settings.libretranslate_url}/translate",
                json={"q": text, "source": source, "target": target, "format": "text"},
            )
            res.raise_for_status()
            return res.json()["translatedText"]
        except (httpx.HTTPError, KeyError):
            # Fallback to MyMemory API
            return await translate_mymemory(client, text, source, target)


async def translate_mymemory(
    client: httpx.AsyncClient, text: str, source: str, target: str
) -> str:
    try:
        res = await client.get(
            "https://api.mymemory.translated.net/get",
            params={"q": text, "langpair": f"{source}|{target}"},
        )
        res.raise_for_status()
        data = res.json()
        return data.get("responseData", {}).get("translatedText", text)
    except httpx.HTTPError:
        return text


async def translate_batch(texts: list[str], source: str, target: str) -> list[str]:
    results = []
    for text in texts:
        translated = await translate(text, source, target)
        results.append(translated)
    return results
